crates/red_knot_python_semantic/src/types.rs

```diff
@@ -2027,6 +2061,33 @@ impl<'db> From<Type<'db>> for Symbol<'db> {
     }
 }
 
+/// Error struct providing information on type(s) that were deemed to be invalid
+/// in a type expression context, and the type we should therefore fallback to
+/// for the problematic type expression.
+#[derive(Debug, PartialEq, Eq)]
+pub struct InvalidTypeExpressionError<'db> {
+    fallback_type: Type<'db>,
+    invalid_expressions: Vec<InvalidTypeExpression>,
```

Comment:
Any reason not to use a boxed array here, since it should be immutable after construction?
```suggestion
    invalid_expressions: Box<[InvalidTypeExpression]>,
```

Reply:
I went for a `Vec` because it's slightly simpler to use (less punctuation to construct it, you don't need the `.into_boxed_slice()` call to convert from the `Vec` into the boxed slice), and I doubt the space savings from using a boxed slice would be significant since it's not like we're storing this array in a cache for any significant amount of time: the array is consumed almost as soon as it's constructed.

However, even better than either a Vec or a boxed slice might be a SmallVec, as then we could avoid allocating in most situations! By far the most common situation is likely to be that only a single diagnostic needs to be emitted